rosidl_generator_rs/resource/action.rs.em

```diff
@@ -0,0 +1,205 @@
+    ACTION_GOAL_SUFFIX,
+    ACTION_RESULT_SERVICE_SUFFIX,
+    ACTION_RESULT_SUFFIX,
+    SERVICE_REQUEST_MESSAGE_SUFFIX,
+    SERVICE_RESPONSE_MESSAGE_SUFFIX,
+)
+
+action_msg_specs = []
+
+for subfolder, action in action_specs:
+    action_msg_specs.append((subfolder, action.goal))
+    action_msg_specs.append((subfolder, action.result))
+    action_msg_specs.append((subfolder, action.feedback))
+    action_msg_specs.append((subfolder, action.feedback_message))
+
+action_srv_specs = []
+
+for subfolder, action in action_specs:
+    action_srv_specs.append((subfolder, action.send_goal_service))
+    action_srv_specs.append((subfolder, action.get_result_service))
+}@
+
+pub mod rmw {
+@{
+TEMPLATE(
+    'msg_rmw.rs.em',
+    package_name=package_name, interface_path=interface_path,
+    msg_specs=action_msg_specs,
+    get_rs_name=get_rs_name, get_rmw_rs_type=get_rmw_rs_type,
+    pre_field_serde=pre_field_serde,
+    get_idiomatic_rs_type=get_idiomatic_rs_type,
+    constant_value_to_rs=constant_value_to_rs)
+
+TEMPLATE(
+    'srv_rmw.rs.em',
+    package_name=package_name, interface_path=interface_path,
+    srv_specs=action_srv_specs,
+    get_rs_name=get_rs_name, get_rmw_rs_type=get_rmw_rs_type,
+    pre_field_serde=pre_field_serde,
+    get_idiomatic_rs_type=get_idiomatic_rs_type,
+    constant_value_to_rs=constant_value_to_rs)
+}@
+}  // mod rmw
+
+@{
+TEMPLATE(
+    'msg_idiomatic.rs.em',
+    package_name=package_name, interface_path=interface_path,
+    msg_specs=action_msg_specs,
+    get_rs_name=get_rs_name, get_rmw_rs_type=get_rmw_rs_type,
+    pre_field_serde=pre_field_serde,
+    get_idiomatic_rs_type=get_idiomatic_rs_type,
+    constant_value_to_rs=constant_value_to_rs)
+}@
+
+@{
+TEMPLATE(
+    'srv_idiomatic.rs.em',
+    package_name=package_name, interface_path=interface_path,
+    srv_specs=action_srv_specs,
+    get_rs_name=get_rs_name, get_rmw_rs_type=get_rmw_rs_type,
+    pre_field_serde=pre_field_serde,
+    get_idiomatic_rs_type=get_idiomatic_rs_type,
+    constant_value_to_rs=constant_value_to_rs)
+}@
+
+@[for subfolder, action_spec in action_specs]
+
+@{
+type_name = action_spec.namespaced_type.name
+}@
+
+#[link(name = "@(package_name)__rosidl_typesupport_c")]
+extern "C" {
+    fn rosidl_typesupport_c__get_action_type_support_handle__@(package_name)__@(subfolder)__@(type_name)() -> *const std::os::raw::c_void;
+}
+
+// Corresponds to @(package_name)__@(subfolder)__@(type_name)
+pub struct @(type_name);
+
+impl rosidl_runtime_rs::Action for @(type_name) {
+  type Goal = crate::@(subfolder)::@(type_name)@(ACTION_GOAL_SUFFIX);
+  type Result = crate::@(subfolder)::@(type_name)@(ACTION_RESULT_SUFFIX);
+  type Feedback = crate::@(subfolder)::@(type_name)@(ACTION_FEEDBACK_SUFFIX);
+
+  fn get_type_support() -> *const std::os::raw::c_void {
+    // SAFETY: No preconditions for this function.
+    unsafe { rosidl_typesupport_c__get_action_type_support_handle__@(package_name)__@(subfolder)__@(type_name)() }
```

Comment:
This is a little confusing to me. Looking at (what I believe to be) the impl for this typesupport function ([link](https://github.com/ros2/rosidl/blob/692b66b0128d1098b5052dc126dfb184a69c1c21/rosidl_typesupport_introspection_c/resource/srv__type_support.c.em#L60))

1. Why are we actually returning a void pointer here? This function is not marked as unsafe, but any handling of the return value may well need to use unsafe code. Could we not return a well formed type, such as `rosidl_service_type_support_t`?

2. This function seems to initialize the request and response members, but I don't actually see where those members are. The `@(type_name)` struct appears empty. How is this working?

Reply:
1. You're right, we could have this return a more specific type than just `c_void`. However, doing so hasn't been necessary yet, and this is following the same pattern as the existing `get_type_support` functions for messages and services. To make this happen, we would have to create a Rust binding for the `rosidl_{message,service,action}_type_support_t` structs, probably in `rosidl_runtime_rs`. I would be inclined to leave this for a separate PR to avoid growing the scope of this one.
2. I'm not sure what you mean here. The `@(type_name)` struct is empty since it's not really meant to be instantiated. It would only be used as a generic `impl Action` argument and to access the specific `Goal`, `Feedback`, and `Result` associated message types.

Reply:
1. Sure, I'm fine leaving this to another PR. Just took a look at our usages ([1](https://github.com/ros2-rust/ros2_rust/blob/main/rclrs/src/client.rs#L95)) of this function historically and we do have `as` casts everywhere. `as` [casts can be problematic](https://doc.rust-lang.org/nomicon/casts.html#safety-of-casting). So it would be nice to eventually not need this. However, maybe there is some other reason I'm missing as to why we return `void*` here.

2. I was referring to the data that function actually mutates in its implementation. I was misunderstanding and assumed that we actually held a handle to that data, but apparently we do not. It is a global variable managed by I guess the rosidl_runtime(?) that will be generated for each service.

```c
static rosidl_service_type_support_t @(function_prefix)__@(spec.srv_name)_service_type_support_handle = {
  0,
  &@(function_prefix)__@(spec.srv_name)_service_members,
  get_service_typesupport_handle_function,
};
```

This is outside the scope of this PR though. Just calling attention to it because we've been bitten by global variables we directly interact with via FFI before, see https://github.com/ros2-rust/ros2_rust/pull/386

Reply:
1. Yes, you're right, it would be good to avoid those `as` casts. I think the only thing preventing that would be defining the appropriate binding type in the `rosidl_runtime_rs` package. I can make an attempt at that in a follow-up PR.
2. The typesupport structs are global statics that are managed by the various typesupport implementations. Some of them seem to handle it a bit more dubiously than others, so that's a good thing to watch out for.